src/expr/src/vector_op/substr.rs

```diff
@@ -36,8 +40,12 @@ pub fn substr_start_for(s: &str, start: i32, count: i32, writer: &mut dyn Write)
     if count < 0 {
         bail!("length in substr should be non-negative: {}", count);
     }
-    let begin = max(start - 1, 0) as usize;
-    let end = min(max(start - 1 + count, 0) as usize, s.len());
+    let start = start.checked_sub(1).ok_or(ExprError::NumericOutOfRange)?;
+    let begin = min(max(start, 0) as usize, s.len());
```

Comment:
Use this instead of `if block` since that is more friendly for branch prediction?
(I'm not very sure)

Reply:
Rust's slicing behavior is kind of weird IMO 🤣 . `&str[s.len()..]` will return an empty slice, while `&s[s.len() + 1..]` will result in panic. So maybe either short circuit or add some comment to explain the correctness when `begin == s.len()`?